### PR TITLE
Fix inference on axes for broadcasting

### LINKED ISSUE #546: Non-inferrability in broadcasting
Noticed in https://github.com/JuliaMath/Interpolations.jl/pull/269:

```julia
using StaticArrays, Test
h = SMatrix{1,1}([1.0])
steps = SVector(1.0)
testinf(h, steps) = h ./ (steps .* steps')

julia> testinf(h, steps)
1×1 SArray{Tuple{1,1},Float64,2,1}:
 1.0

julia> @inferred testinf(h, steps)
ERROR: return type SArray{Tuple{1,1},Float64,2,1} does not match inferred return type Any
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] top-level scope at none:0
```

### PR DESCRIPTION
**Pull request by timholy**
Fixes #546

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Does this fix https://github.com/JuliaArrays/StaticArrays.jl/issues/540 ?

**Comment by timholy** [comment]
Nope. Also doesn't fix #537 though it would seem that things here should at least make that easier. Finally, doesn't fix #485.

**Comment by timholy** [comment]
See also https://github.com/JuliaLang/julia/pull/30074.

File: src/SOneTo.jl [hunk 1]
```diff
@@ -58,3 +58,6 @@ end
 Base.@pure function Base.checkindex(::Type{Bool}, ::SOneTo{n1}, ::SOneTo{n2}) where {n1, n2}
     return n1::Int >= n2::Int
 end
+
+Base.promote_rule(a::Type{Base.OneTo{T}}, ::Type{SOneTo{n}}) where {T,n} =
+    Base.OneTo{promote_type(T, Int)}
```
Comment (timholy) [inline src/SOneTo.jl]: Hmm, this is not strictly necessary for the fix (it's just the first thing I tried). I could get rid of this if you have reservations about it. (One also has to be rid of the tests this adds to `core.jl`.)

Comment (andyferris) [reply src/SOneTo.jl]: I think this is good.

**Pull request merged**